external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReader.scala

```diff
@@ -419,6 +416,19 @@ private[kafka010] class KafkaOffsetReader(
     stopConsumer()
     _consumer = null  // will automatically get reinitialized again
   }
+
+  private def getPartitions(): ju.Set[TopicPartition] = {
+    var partitions = Set.empty[TopicPartition].asJava
+    val startTimeMs = System.currentTimeMillis()
+    while (partitions.isEmpty && System.currentTimeMillis() - startTimeMs < pollTimeoutMs) {
+      // Poll to get the latest assigned partitions
+      consumer.poll(jt.Duration.ofMillis(100))
```

Comment:
I'm sorry for back and forth on this, but it would be better if we can add timeout only when it's retrying. So try once with no delay, and retry with some timeouts (effectively sleep) only if it fails. 

This would avoid introducing additional timeout when the metadata is available, but still adding some delay between retrying. WDYT?

Reply:
If assignment is available and the consumer is able to poll then `consumer.poll(jt.Duration.ofMillis(100))` call returns immediately (not mentioning the time required to poll) so not yet understand why would it be better.

Based on your suggestion I've concluded the following implementation (correct me if I misunderstood):
```
var firstFetch = false
...
if (!firstFetch) {
  consumer.poll(jt.Duration.Zero)
  firstFetch = true
} else {
  consumer.poll(jt.Duration.ofMillis(100))
}
...
```
Does it worth such complication to spare 100ms?

Reply:
No you just need to start the method with below:

```
consumer.poll(jt.Duration.Zero)
var partitions = consumer.assignment()
```

instead of initializing partitions as empty set.

If `consumer.poll(0)` works and partitions are filled with assignment, loop will not be executed. Does it work for you?

Reply:
Aha, I see your point. What you're suggesting is correct.
On the other hand still don't understand the point. If assignment is available `consumer.poll(jt.Duration.Zero)` should return the same speed just like `consumer.poll(jt.Duration.ofMillis(100))`.

Reply:
Please correct me if I'm missing here. (Not an expert of Kafka, may miss some details.) 

It may return after 100ms if there's no record to consume even metadata is ready. Here we only need metadata but once we call poll, the request is bound to the records. 

To be clear, we would also like to call `poll(0)` with explicitly putting sleep (to avoid coupling with records), but it would be also OK to let `consumer.poll` wait instead.